### ISSUE TITLE
Support for NGSIv2

### ISSUE THREAD

**Issue by dmoranj** [telefonicaid/iotagent-node-lib#527]
Currently, all the IoTAgents communicate with the Context Broker using NGSIv1 for all the communications (both for context provisioning and context management). A migration to NGSIv2 is needed, as some behaviors are starting to diverge (mainly the use of special types in NGSIv2).

**Comment by dcalvoalonso** [comment]
We propose to create a first functional version to support the [Release Candidate version of NGSIv2](http://telefonicaid.github.io/fiware-orion/api/v2/stable/). This first version should be a Minimum Viable Product (MVP) that provides a relevant set of NGSIv2 functionalities.

The following modifications should be added:
* Communication with Context Broker considers new URLs and payloads (following RESTful principles).
  * Entity Creation.
  * Query entity. Including options to retrieve keyValues, values for a list of attributes or a single attribute.
  * Text-based attribute value set/get
  * Getting all entities and filtering.
  * Update entity attributes and a single attribute.
  * Creation of subscriptions.
  * Pause of subscriptions.
  * Delete subscriptions.
* Native JSON datatypes.
* Datetime support based on ISO ISO8601 format.
* Communication with Context Broker considers new error responses.
* Unit tests and examples cover completely the new modifications (100% of code coverage).

One important question to be answered is if we should maintain compatibility with NGSIv1 in the IoT Agents.  If we do so, we should introduce an option in the configuration to choose which version of NGSI to use. Obviously, the complexity of the resulting code will be higher.

The next step would be the migration of [iot.IoTagent-UL](https://github.com/Fiware/iot.IoTagent-UL), [iot.IoTagent-JSON](https://github.com/Fiware/iot.IoTagent-JSON) and [iot.IoTagent-LWM2M](https://github.com/Fiware/iot.IoTagent-LWM2M).

**Comment by fgalan** [comment]
@dcalvoalonso thank you for your contribution to issue comments!

Please, take into account the following feedback:

* With regards to NGSIv2 operations, note that IOTAs use only a subset of the API. In particular, create/pause/delete subscriptions and get all entities and filtering are not used by ITOAs. I have also doubts with regards to text-based set/get... probablly it is also unneeded. The basic NGSI operations IOTA uses are:
  * Registration to CB on device provision time. Note that NGSIv2 registrations has not been yet implemented in CB (in fact, that piece of work is starting). However, the specification has been defined in the [lastest flavour of the spec](http://telefonicaid.github.io/fiware-orion/api/v2/latest/), eventually to be consolidated at new RC version. Thus you could use it to guide your implemention of the client part at IOTA side and not block until CB implementation gets ready.
  * Send updates to CB (for active attributes).
  * Acts as Context Provider for queries/updates coming from CB (for passive attributes and commands). Note that NGSIv2 based forwarding has not been yet implemented at Orion. Different from NGSIv2 registrations (which are in the spec), in this case, NGSIv2 forwarding has not been even defined at spec level. @jmcanterafonseca could you provide some hint about the status of this work, please?
* Regarding backward compability, I think introducing an option in config.js to switch between NGSIv1 and NGSIv2 would be a good idea. It would allow to quickly go back to NGSIv1 if some problem is found in production with NGSIv2, which is very important from backward compability point of view. Once the NGSIv2 version gets tested (during some months) and we are confident with it, the old NGSIv1 code could be removed. If the code is well moduled, it shouldn't be too complex to include such switch, just an "if" in the right place to route the execution of code either to the NGSIv1 and NGSIv2 part. Note the big part of the implementation (so, the complexity), i.e to write the NGSIv2 code and write the NGSIv2, is orthogonal to having the switch or not. What do you think?
* With regards to migrating particular IOTA instances using this library (IOTA-UL, etc.) probably you don't need to do any migration (expect updatinng the library version at packages.json, I guess) as all the NGSI-stuff is included in the library. However, let's see when the moment comes... :)

**Comment by dcalvoalonso** [comment]
Thanks for the prompt reply @fgalan !! :)

* I agree with you, only the specific functionalities of NGSIv2 that are needed by IOTA should be implemented. I will pay special attention to your recommendations.
* Maintaining compatibility with NGSIv1 is a good point in order to do a smooth and reliable transition. Let's follow this approach.
* I have already done some research about how to migrate each particular IOTA and I guess that as you explain, changes are not really complicated. In most cases, just to modify unit tests.

**Comment by fgalan** [comment]
Updates to CB using NGSIv2 implemented in PR https://github.com/telefonicaid/iotagent-node-lib/pull/556

**Comment by dcalvoalonso** [comment]
I have already implemented the needed modifications in the following plugins:
* Timestamp Compression plugin
* Attribute Alias plugin
* Event plugin
* Timestamp Processing Plugin
* Expression Translation plugin

If you want, you can check the changes in https://github.com/dcalvoalonso/iotagent-node-lib/tree/task/pluginsNgsi2. I have followed the approach I used yesterday to send active attributes. I will not submit the PR until PR #556 to avoid having too much open PRs in the repository.

Regarding the rest of plugins:
* Bidirectionality plugin: I have discovered that this plugin does require that IOTAs implements subscriptions, as you can see in the [documentation](https://github.com/telefonicaid/iotagent-node-lib#bidirectionality-plugin-bidirectional). **What do you think? If this plugin is going to be valid, we should implement finally NGSIv2 subscriptions.**
* Multientity plugin: This is a bit more complex than the previous ones since in NGSIv2, there is an option to update several entities with a single POST operation. I will do the implementation at the end because I guess that some additional changes in ngsiService will be needed.

**Comment by fgalan** [comment]
Yes, let's merge first #556 (which in sequence requires solving the travis problem) before making next move in this line.

**Comment by fgalan** [comment]
With regards to:

> What do you think? If this plugin is going to be valid, we should implement finally NGSIv2 subscriptions.

I'm not sure if documentation is misaligned here... I mean, are NGSIv1 subscription and notification handlers actually implemented in the code? Could you point to the part of the code where NGSIv1 subscription payload is composed and send and the handler were notification payload is recieved and consumed, please?

**Comment by dcalvoalonso** [comment]
Yes, they are actually implemented for NGSIv1. Let's going to see how they are used with bidirectional data plugin:
1. In [Bidirectional data plugin test]https://github.com/telefonicaid/iotagent-node-lib/blob/master/test/unit/plugins/bidirectional-plugin_test.js
 * In lines 83-88, you can see that using this plugin implies inserting new middleware functions for device provision, configuration provision and **notifications**.
2. In [Bidirectional data plugin ](https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/plugins/bidirectionalData.js)
 * In line 223, function handleDeviceProvision. As you can see in line 230, it calls to sendSubscriptions method. This method is defined in line 100 and calls to subscribe method of (subscriptionService)[https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/services/ngsi/subscriptionService.js]. In line 94 of this file, you can see how the payload is built and sent to the CB and in line 109, that providerUrl of the configuration file is used to create the reference key and therefore to receive the notifications from the CB.
 * In line 235 the result of this call is used to update the device that will be finally stored in the IOTARegistry.
3. In [context server](https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/services/northBound/contextServer.js):
 * In line 779, the endpoint to receive notifications from the CB is created. As you can see, it is also possible to insert new middleware functions.
4. Back in [Bidirectional data plugin ](https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/plugins/bidirectionalData.js), in line 250 you can see the method to handle incoming notifications.

**Comment by fgalan** [comment]
Timestamp compression plugin addressed in PR https://github.com/dcalvoalonso/iotagent-node-lib/pull/1

**Comment by fgalan** [comment]
Useful information regarding pluging usage in thre 3 main IOTAs using the libray (raised at https://github.com/dcalvoalonso/iotagent-node-lib/pull/1#discussion_r151154358):

* iotagent-json:
   * It uses all the plugins in https://github.com/telefonicaid/iotagent-node-lib#provided-plugins
* iotagent-ul
   * It uses all the others plugins in https://github.com/telefonicaid/iotagent-node-lib#provided-plugins except Time Compression plugin
* lightweightm2m-iotagent
   * It doesn't uses any plugin in https://github.com/telefonicaid/iotagent-node-lib#provided-plugins at all

**Comment by fgalan** [comment]
Timestamp Processing plugin addressed in PR https://github.com/dcalvoalonso/iotagent-node-lib/pull/2

(3 more to go, in the basic set)

**Comment by fgalan** [comment]
Event, Alias and Expression plugin addressed in PR https://github.com/dcalvoalonso/iotagent-node-lib/pull/3

Regarding the rest of the plugins:

> Bidirectionality plugin: I have discovered that this plugin does require that IOTAs implements subscriptions, as you can see in the documentation. What do you think? If this plugin is going to be valid, we should implement finally NGSIv2 subscriptions.

If this plugin uses NGSIv2 subscriptions, then it also should be migrated. Note NGSIv2 allows several "flavours" for notification formats (check "Notification Messages" section at http://telefonicaid.github.io/fiware-orion/api/v2/stable/. I'd suggest to use `normalized` format, as it is the more complete.

> Multientity plugin: This is a bit more complex than the previous ones since in NGSIv2, there is an option to update several entities with a single POST operation. I will do the implementation at the end because I guess that some additional changes in ngsiService will be needed.

Makes sense.

**Comment by dcalvoalonso** [comment]
New PR for bidirectionality plugin and subscriptions: https://github.com/dcalvoalonso/iotagent-node-lib/pull/4

**Comment by dcalvoalonso** [comment]
Regarding multitentity plugin, I have studied how this works for NGSIv1:
1. The objetive of this plugin is: 

> Allows the devices provisioned in the IoTAgent to map their attributes to more than one entity, declaring the target entity through the Configuration or Device provisioning APIs.

2. In line 252 of function sendUpdateValue of [ngsiService.js](https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/services/ngsi/ngsiService.js):

```
        apply(applyMiddlewares, updateMiddleware, payload, typeInformation)

```

3.  In order to use the plugin we just need to do:

```
                iotAgentLib.addUpdateMiddleware(iotAgentLib.dataPlugins.multiEntity.update);

```

This basically adds a new middleware function that will be called in the code of step 2.

4. What the plugin does is just add new context entities to the request. In line 106 of [multiEntity.js](https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/plugins/multiEntity.js): it adds new context entities to be updated.

The problem with NGSIv2 is that the operation to update attributes does only support doing it for a single entity (https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/user/walkthrough_apiv2.md#update-entity).

Therefore, from my point of view, the only option we have is to rely on [batch operations](https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/user/walkthrough_apiv2.md#batch-operations).

What do you think? Do you see other option?

**Comment by fgalan** [comment]
If I understanding correctly you mean that in the current version of the plugin (based in NGSIv1) the update of the multiple entieies is done in just one request towards CB.

In order to have a better idea of the situtation and provide better feedback, could you provide an example of such request (verb + URL + payload), pls?

**Comment by dcalvoalonso** [comment]
For NGSIv1, the plugin sends the following request to the CB in order to update multiple entities:

```
POST /v1/updateContext
{
  "contextElements": [
    {
      "type": "WeatherStation",
      "isPattern": "false",
      "id": "ws4",
      "attributes": [
        {
          "name": "pressure",
          "type": "Hgmm",
          "value": "52"
        }
      ]
    },
    {
      "type": "Higrometer",
      "isPattern": "false",
      "id": "Higro2000",
      "attributes": [
        {
          "name": "humidity",
          "type": "Percentage",
          "value": "12"
        }
      ]
    }

  ],
  "updateAction": "UPDATE"
}

```

After having read the documentation for the NGSIv2 API of CB, I understand that the only possibility to do this kind of multiple updates is to use batch operations, e.g.:

```
POST POST /v2/op/update

{
  "actionType": "UPDATE",
  "entities": [
    {
      "type": "WeatherStation",
      "id": "ws4",
      "pressure": {
        "value": "52",
        "type": "Hgmm"
      }
    },
    {
      "type": "Higrometer",
      "id": "Higro2000",
      "humidity": {
        "value": "12",
        "type": "Percentage"
      }
    }
  ]
}
```

**Comment by fgalan** [comment]
`POST /v2/op/update` in the equivalent in NGSIv2 to `POST /v1/updateContext` in NGSIv1. Thus, it makes sense to use that one.

**Comment by dcalvoalonso** [comment]
New PR for multientity plugin: https://github.com/dcalvoalonso/iotagent-node-lib/pull/5

**Comment by dcalvoalonso** [comment]
After having received @fgalan LGTM to PR https://github.com/dcalvoalonso/iotagent-node-lib/pull/5, I would like to do a quick summary of the status of this task. We have already implemented and unit tested the following features:
* Send updates to CB (for active attributes) https://github.com/telefonicaid/iotagent-node-lib/pull/556
* Data plugins:
 * Timestamp Compression plugin https://github.com/dcalvoalonso/iotagent-node-lib/pull/1
 * Attribute Alias plugin https://github.com/dcalvoalonso/iotagent-node-lib/pull/3
 * Event plugin https://github.com/dcalvoalonso/iotagent-node-lib/pull/3
 * Timestamp Processing plugin https://github.com/dcalvoalonso/iotagent-node-lib/pull/2
 * Expression Translation plugin https://github.com/dcalvoalonso/iotagent-node-lib/pull/3
 * Bidirectionality plugin https://github.com/dcalvoalonso/iotagent-node-lib/pull/4
 * Multientity plugin https://github.com/dcalvoalonso/iotagent-node-lib/pull/5
* Create and delete subscriptions in CB https://github.com/dcalvoalonso/iotagent-node-lib/pull/4

All these features have been merged in https://github.com/dcalvoalonso/iotagent-node-lib/tree/task/activeAttsNgsi2.

As it can be seen in https://github.com/telefonicaid/iotagent-node-lib/pull/556, there are no conflicts with master branch, unit and lint tests are passed.

A final step must be completed before approving the PR #556: integration tests against Orion CB.

**Comment by dcalvoalonso** [comment]
In addition, the following functionalities must be migrated to NGSIv2:
* Device provisioning
* Lazy attributes
* Commands

However, @fgalan explained in https://github.com/telefonicaid/iotagent-node-lib/issues/527#issuecomment-337565075 that:

> NGSIv2 registrations has not been yet implemented in CB.

Then although we could implement the entity creation in NGSIv2, we agreed in that it would be better not to do a partial development.

And:

> NGSIv2 forwarding has not been even defined at spec level.

An issue has been recently opened in the CB repository to progress on this: https://github.com/telefonicaid/fiware-orion/issues/3068

**Comment by fgalan** [comment]
> A final step must be completed before approving the PR #556: integration tests against Orion CB.

Doubt: can the integration tests be done just based in the work done in iotagents library? Or the issues in the IOTAs themselves

* https://github.com/telefonicaid/lightweightm2m-iotagent/issues/104
* https://github.com/telefonicaid/iotagent-ul/issues/233
* https://github.com/telefonicaid/iotagent-json/issues/250

should be completed before?

**Comment by dcalvoalonso** [comment]
We have most of the migration to NGSIv2 for each one of the IOTAs ready and our intention is to start submitting PRs by the end of this week or the beginning of the next one. However, I am going to analyze if it is possible to run some integration tests with the iotagent-node-lib. I will let you know my conclusions.

**Comment by fgalan** [comment]
PR #556 implementing active attributes sending using NGSIv2 (across all the plugins where this functionality plays a role) has been merged.

**Comment by dcalvoalonso** [comment]
@fgalan , could you update which is the status of the pending NGSIv2 implementations in the CB (i.e. entity provisioning and  forwarded request processing)?

**Comment by fgalan** [comment]
With regards with entity provision, entity create/update operation are available in CB time ago. Registration operation get implemented recently and are available in the last Orion version (1.13.0) and also in master branch, of course.

With regards to query/update forwarding based on registrations, I'm afraid this is not even specified. See https://github.com/telefonicaid/fiware-orion/issues/3068 for details. As far as I remember @jmcanterafonseca got the AP of making a first draft proposal.

**Comment by fgalan** [comment]
NGSIv2-based device provisioning/de-provisioning implemented in PR https://github.com/telefonicaid/iotagent-node-lib/pull/615

**Comment by dcalvoalonso** [comment]
After the merge of:
- iotagent-node-lib: https://github.com/telefonicaid/iotagent-node-lib/pull/615
- iotagent-json: https://github.com/telefonicaid/iotagent-json/pull/284
- iotagent-ul: https://github.com/telefonicaid/iotagent-ul/pull/260
- lightweightm2m-iotagent: https://github.com/telefonicaid/lightweightm2m-iotagent/pull/126

Entity provisioning using NGSIv2 is implemented.

The only missing point is forwarded request processing, which is still in the process of being specified in NGSIv2.

**Comment by fgalan** [comment]
> The only missing point is forwarded request processing, which is still in the process of being specified in NGSIv2.

Implemented by PR https://github.com/telefonicaid/iotagent-node-lib/pull/762, based in a simplified apporach described in https://github.com/telefonicaid/fiware-orion/issues/3068#issuecomment-468584414

**Comment by fgalan** [comment]
PR #762 was overpassed by PR https://github.com/telefonicaid/iotagent-node-lib/pull/813.

Nothing is pending in the issue. Closing.
